File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
def solution(numbers):
    answer = 0
    visited = [False] * 7
    nums = set()
    
    def perm(numbers, num, depth):
        if depth > len(numbers):
            return
        
        for i in range(len(numbers)):
            if not visited[i]:
                visited[i] = True
                nums.add(int(f'{num}{numbers[i]}'))
                perm(numbers, f'{num}{numbers[i]}', depth + 1)
                visited[i] = False
    
    perm(numbers, "", 0)
    
    for n in nums:
        if n == 0 or n == 1:
            continue
        is_prime = True
        for i in range(2, int(n ** 0.5) + 1):
            if n % i == 0:
                is_prime = False
        if is_prime:
            answer += 1
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    private static Set<Integer> results = new HashSet();
    
    private static void permutation(String buffer, String numbers, boolean[] visited) {
        for (int j = 0; j < numbers.length(); j++) {
            if (!visited[j]) {
                visited[j] = true;
                String merged = buffer + numbers.charAt(j);
                results.add(Integer.valueOf(merged));
                permutation(merged, numbers, Arrays.copyOf(visited, visited.length));
                visited[j] = false;
            }
        }
    }
    
    private static boolean isPrimeNumber(int num) {
        if (num < 2) {
            return false;
        }
        for (int i = 2; i <= Math.sqrt(num); i++) {
            if (num % i == 0)
                return false;
        }
        return true;
    }
    public int solution(String numbers) {
        int answer = 0;
        // 나올 수 있는 숫자의 모든 경우를 구한다
        permutation("", numbers, new boolean[numbers.length()]);
        // 각 경우가 소수면 answer++
        for (Integer s: results) {
            if (isPrimeNumber(s))
                answer++;
        }
        return answer;
    }
}